internal/action/render.go

```diff
@@ -0,0 +1,235 @@
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/operator-framework/operator-registry/internal/declcfg"
+	"github.com/operator-framework/operator-registry/internal/property"
+	"github.com/operator-framework/operator-registry/pkg/containertools"
+	"github.com/operator-framework/operator-registry/pkg/image"
+	"github.com/operator-framework/operator-registry/pkg/image/containerdregistry"
+	"github.com/operator-framework/operator-registry/pkg/lib/bundle"
+	"github.com/operator-framework/operator-registry/pkg/registry"
+	"github.com/operator-framework/operator-registry/pkg/sqlite"
+)
+
+type Render struct {
+	Refs     []string
+	Registry image.Registry
+}
+
+func nullLogger() *logrus.Entry {
+	logger := logrus.New()
+	logger.SetOutput(ioutil.Discard)
+	return logrus.NewEntry(logger)
+}
+
+func (r Render) Run(ctx context.Context) (*declcfg.DeclarativeConfig, error) {
+	if r.Registry == nil {
+		reg, err := r.createRegistry()
+		if err != nil {
+			return nil, fmt.Errorf("create registry: %v", err)
+		}
+		defer reg.Destroy()
+		r.Registry = reg
+	}
+
+	var cfgs []declcfg.DeclarativeConfig
+	for _, ref := range r.Refs {
+		var (
+			cfg *declcfg.DeclarativeConfig
+			err error
+		)
+		if stat, serr := os.Stat(ref); serr == nil && stat.IsDir() {
+			cfg, err = declcfg.LoadDir(ref)
+		} else {
+			cfg, err = r.imageToDeclcfg(ctx, ref)
+		}
+		if err != nil {
+			return nil, fmt.Errorf("render reference %q: %v", ref, err)
+		}
+		renderBundleObjects(cfg)
+		cfgs = append(cfgs, *cfg)
+	}
+
+	return combineConfigs(cfgs), nil
+}
+
+func (r Render) createRegistry() (*containerdregistry.Registry, error) {
+	cacheDir, err := os.MkdirTemp("", "render-registry-")
+	if err != nil {
+		return nil, fmt.Errorf("create tempdir: %v", err)
+	}
+
+	reg, err := containerdregistry.NewRegistry(
+		containerdregistry.WithCacheDir(cacheDir),
+		containerdregistry.WithLog(nullLogger()),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return reg, nil
+}
+
+func (r Render) imageToDeclcfg(ctx context.Context, imageRef string) (*declcfg.DeclarativeConfig, error) {
+	ref := image.SimpleReference(imageRef)
+	if err := r.Registry.Pull(ctx, ref); err != nil {
+		return nil, err
+	}
+	labels, err := r.Registry.Labels(ctx, ref)
+	if err != nil {
+		return nil, err
+	}
+	tmpDir, err := ioutil.TempDir("", "render-unpack-")
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(tmpDir)
+	if err := r.Registry.Unpack(ctx, ref, tmpDir); err != nil {
+		return nil, err
+	}
+
+	var cfg *declcfg.DeclarativeConfig
+	if dbFile, ok := labels[containertools.DbLocationLabel]; ok {
+		cfg, err = sqliteToDeclcfg(ctx, filepath.Join(tmpDir, dbFile))
+		if err != nil {
+			return nil, err
+		}
+	} else if configsDir, ok := labels["operators.operatorframework.io.index.configs.v1"]; ok {
+		// TODO(joelanford): Make a constant for above configs location label
```

Comment:
+1

Reply:
It isn't clear to me _where_ to put this constant, which is why I added the TODO comment? Any suggestions?

Reply:
(Maybe this gets rolled into the other follow-up work that will likely happen around exporting some of the DC-related code?)